server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java

```diff
@@ -0,0 +1,266 @@
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.BitSet;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+final class RecoverySourcePruneMergePolicy extends OneMergeWrappingMergePolicy {
+    RecoverySourcePruneMergePolicy(String recoverySourceField, Supplier<Query> retentionPolicySupplier, MergePolicy in) {
+        super(in, toWrap -> new OneMerge(toWrap.segments) {
+            @Override
+            public CodecReader wrapForMerge(CodecReader reader) throws IOException {
+                CodecReader wrapped = toWrap.wrapForMerge(reader);
+                NumericDocValues recovery_source = wrapped.getNumericDocValues(recoverySourceField);
+                if (recovery_source == null || recovery_source.nextDoc() == DocIdSetIterator.NO_MORE_DOCS) {
+                    return wrapped;
+                }
+                return wrapReader(recoverySourceField, wrapped, retentionPolicySupplier);
+            }
+        });
+
+    }
+
+    // pkg private for testing
+    static CodecReader wrapReader(String recoverySourceField, CodecReader reader, Supplier<Query> retentionPolicySupplier)
+        throws IOException {
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        builder.add(new DocValuesFieldExistsQuery(recoverySourceField), BooleanClause.Occur.FILTER);
+        builder.add(retentionPolicySupplier.get(), BooleanClause.Occur.FILTER);
+        IndexSearcher s = new IndexSearcher(reader);
+        s.setQueryCache(null);
+        Weight weight = s.createWeight(builder.build(), false, 1.0f);
+        Scorer scorer = weight.scorer(reader.getContext());
+
+        if (scorer != null) {
+            BitSet sourceToDrop = BitSet.of(scorer.iterator(), reader.maxDoc());
+            return new SourcePruningFilterCodecReader(recoverySourceField, reader, sourceToDrop);
+        } else {
+            return reader;
+        }
+    }
+
+    private static class SourcePruningFilterCodecReader extends FilterCodecReader {
+        private final BitSet sourceToDrop;
+        private final String recoverySourceField;
+
+        SourcePruningFilterCodecReader(String recoverySourceField, CodecReader reader, BitSet sourceToDrop) {
+            super(reader);
+            this.recoverySourceField = recoverySourceField;
+            this.sourceToDrop = sourceToDrop;
+        }
+
+        @Override
+        public DocValuesProducer getDocValuesReader() {
+            DocValuesProducer docValuesReader = super.getDocValuesReader();
+            return new FilterDocValuesProducer(docValuesReader) {
+                @Override
+                public NumericDocValues getNumeric(FieldInfo field) throws IOException {
+                    NumericDocValues numeric = super.getNumeric(field);
+                    if (recoverySourceField.equals(field.name)) {
+                        return new FilterNumericDocValues(numeric) {
```

Comment:
maybe leave comments about why we don't need to check whether `numeric` and `docValuesReader` are `null`